models/mli/fairness/model_fairxgb.py

```diff
@@ -0,0 +1,506 @@
+            os.mkdir(tmp_folder)
+        except FileExistsError:
+            # We should never be here since temp dir name is expected to be unique
+            loggerwarning(logger, "FAIRXGB temp folder already exists")
+            tmp_folder = os.path.join(self.context.experiment_tmp_dir, "%s_FAIRXGB_model_folder" % uuid.uuid4())
+            os.mkdir(tmp_folder)
+        except:
+            # Revert to temporary file path
+            tmp_folder = os.path.join(user_dir(), "%s_FAIRXGB_model_folder" % uuid.uuid4())
+            os.mkdir(tmp_folder)
+
+        loggerinfo(logger, "FAIRXGB temp folder {}".format(tmp_folder))
+        return tmp_folder
+
+
+    def fit(self, X, y, sample_weight=None, eval_set=None, sample_weight_eval_set=None, **kwargs):
+             
+        # Specify these parameters for the dataset
+        # Also set feature engineering effort to 0
+        # under the features section of expert settings
+        ########################
+        # Specify the protected column
+        # Must be numeric
+        self.protected_name = "black"
+        # Specify the level of the protected group in the protected column
+        # Must be numeric
+        self.protected_label = 1
+        # Specify the target level considered to be a positive outcome
+        # Must be encoded as 0/1
+        self.positive_target = 0
+        # Set minimum mean protected ratio desired  
+        # (mean protected ratio = mean predictions for the protected group/mean predictions for all other groups)
+        #
+        # Try tuning this to values at or a little above
+        # the mean of the positive target for the protected group
+        # divided by the mean of the positive target for the unprotected group
+        # If it's set too large, the accuracy will be poor, so there
+        # is a limit to the debiasing that can be obtained.
+        self.mean_protected_prediction_ratio_minimum = 0.95
+        ########################
+        
+        orig_cols = list(X.names)
+              
+        import pandas as pd
+        import numpy as np
+        from sklearn.preprocessing import OneHotEncoder
+        from collections import Counter
+        import xgboost as xgb
+
+              
+        # Get the logger if it exists
+        logger = None
+        if self.context and self.context.experiment_id:
+            logger = make_experiment_logger(experiment_id=self.context.experiment_id,
+                                                 tmp_dir=self.context.tmp_dir,
+                                                 experiment_tmp_dir=self.context.experiment_tmp_dir)
+              
+        # Current mu value
+        mu = self.params["mu"]
+        
+                
+        def fair_metric(predt: np.ndarray, dtrain: xgb.DMatrix):
+            ''' FairXGB Error Metric'''
+
+            # Find the right protected group vector
+            if len(predt) == len(protected_train):
```

Comment:
what is `predt` in this case? the protected column?

Reply:
It's the predictions.  I should comment about that.

Reply:
Thank you